### PR TITLE
[SPARK-48267][SS] Regression e2e test with SPARK-47305

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to add a regression test (e2e) with SPARK-47305.

As of commit cae2248bc13 (pre-Spark 4.0), the query in new unit test is represented as below logical plans:

> Batch 0

>> analyzed plan

```
WriteToMicroBatchDataSource MemorySink, 5067923b-e1d0-484c-914c-b111c9e60aac, Append, 0
+- Project [value#1]
   +- Join Inner, (cast(code#5 as bigint) = ref_code#14L)
      :- Union false, false
      :  :- Project [value#1, 1 AS code#5]
      :  :  +- StreamingDataSourceV2ScanRelation[value#1] MemoryStreamDataSource
      :  +- Project [value#3, cast(code#9 as int) AS code#16]
      :     +- Project [value#3, null AS code#9]
      :        +- LocalRelation <empty>, [value#3]
      +- Project [id#12L AS ref_code#14L]
         +- Range (1, 5, step=1, splits=Some(2))
```

>> optimized plan

```
WriteToDataSourceV2 MicroBatchWrite[epoch: 0, writer: ...]
+- Join Inner
   :- StreamingDataSourceV2ScanRelation[value#1] MemoryStreamDataSource
   +- Project
      +- Filter (1 = id#12L)
         +- Range (1, 5, step=1, splits=Some(2))
```

> Batch 1

>> analyzed plan

```
WriteToMicroBatchDataSource MemorySink, d1c8be66-88e7-437a-9f25-6b87db8efe17, Append, 1
+- Project [value#1]
   +- Join Inner, (cast(code#5 as bigint) = ref_code#14L)
      :- Union false, false
      :  :- Project [value#1, 1 AS code#5]
      :  :  +- LocalRelation <empty>, [value#1]
      :  +- Project [value#3, cast(code#9 as int) AS code#16]
      :     +- Project [value#3, null AS code#9]
      :        +- StreamingDataSourceV2ScanRelation[value#3] MemoryStreamDataSource
      +- Project [id#12L AS ref_code#14L]
         +- Range (1, 5, step=1, splits=Some(2))
```

>> optimized plan

```
WriteToDataSourceV2 MicroBatchWrite[epoch: 1, writer: ...]
+- Join Inner
   :- StreamingDataSourceV2ScanRelation[value#3] MemoryStreamDataSource
   +- LocalRelation <empty>
```

Notice the difference in optimized plan between batch 0 and batch 1. In optimized plan for batch 1, the batch side is pruned out, which goes with the path of PruneFilters. The sequence of optimization is, 

1) left stream side is collapsed with empty local relation
2) union is replaced with subtree for right stream side as left stream side is simply an empty local relation
3) the value of 'code' column is now known to be 'null' and it's propagated to the join criteria (`null = ref_code`)
4) join criteria is extracted out from join, and being pushed to the batch side
5) the value of 'ref_code' column can never be null, hence the filter is optimized as `filter false`
6) `filter false` triggers PruneFilters (where we fix a bug in SPARK-47305)

Before SPARK-47305, a new empty local relation was incorrectly marked as streaming.

NOTE: I intentionally didn't put the detail like above as code comment, as optimization result is subject to change for Spark versions.

### Why are the changes needed?

In the PR of SPARK-47305 we only added an unit test to verify the fix, but it wasn't e2e about the workload we encountered an issue. Given the complexity of QO, it'd be ideal to put an e2e reproducer (despite simplified) as regression test.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New UT.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @cloud-fan @viirya PTAL, thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryOptimizationCorrectnessSuite.scala [hunk 1]
```diff
@@ -416,4 +416,39 @@ class StreamingQueryOptimizationCorrectnessSuite extends StreamTest {
       )
     }
   }
+
+  test("SPARK-48267: regression test, stream-stream union followed by stream-batch join") {
+    withTempDir { dir =>
+      val input1 = MemoryStream[Int]
+      val input2 = MemoryStream[Int]
+
+      val df1 = input1.toDF().withColumn("code", lit(1))
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryOptimizationCorrectnessSuite.scala:425]: Given the sequence of optimizations in reproducer, it wouldn't happen if the union wasn't eliminated, as the value of 'code' won't be known during optimization phase. That said, if we call observe() or withWatermark() or etc in these streaming DF, the bug wouldn't be triggered.

**Comment by HeartSaVioR** [comment]
Looks like the behavior of PruneFilters is somewhat interesting, with combination of filter pushdown... If the filter can push down more aggressively, the scope of subtree PruneFilters will prune is going to be "smaller".

E.g. 

```
      val df = input1.toDF()
        .observe("myEvent", count(lit(1)).as("rows"))
        .filter(expr("false"))
```

to

```
== Analyzed Logical Plan ==
WriteToMicroBatchDataSource MemorySink, 5afa5bf2-9757-4e90-a604-bb2c4438ebec, Update, 0
+- Filter false
   +- CollectMetrics myEvent, [count(1) AS rows#3L], 0
      +- StreamingDataSourceV2ScanRelation[value#1] MemoryStreamDataSource

== Optimized Logical Plan ==
WriteToDataSourceV2 MicroBatchWrite[epoch: 0, writer: org.apache.spark.sql.execution.streaming.sources.MemoryStreamingWrite@64b874f0]
+- LocalRelation <empty>, [value#1]
```

vs

```
      val df = input1.toDF()
        .withColumn("eventTime", timestamp_seconds($"value"))
        .withWatermark("eventTime", "0 second")
        .filter(expr("false"))
```

to

```
== Analyzed Logical Plan ==
WriteToMicroBatchDataSource MemorySink, 371ac05f-8c25-43b1-b82a-ed11e9a16c29, Update, 0
+- Filter false
   +- EventTimeWatermark eventTime#3: timestamp, 0 seconds
      +- Project [value#1, timestamp_seconds(value#1) AS eventTime#3]
         +- StreamingDataSourceV2ScanRelation[value#1] MemoryStreamDataSource

== Optimized Logical Plan ==
WriteToDataSourceV2 MicroBatchWrite[epoch: 0, writer: org.apache.spark.sql.execution.streaming.sources.MemoryStreamingWrite@4accccf7]
+- EventTimeWatermark eventTime#3: timestamp, 0 seconds
   +- LocalRelation <empty>, [value#1, eventTime#3]
```

For former, filter can't be pushed down under CollectMetrics, and PruneFilters takes effect and replace both CollectMetrics and Relation with an empty relation.

For latter, EventTimeWatermark supports pushdown which predicate isn't comparing with event time column. This case filter is pushed down under EventTimeWatermark, and PruneFilters takes effect and replace Relation, but EventTimeWatermark is left as it is.

**Review by viirya — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
Thanks @viirya for quick reviewing! Merging to master.

**Comment by HeartSaVioR** [comment]
Also cherry-picked to 3.5 as well as it's a clean cherry-pick.

**Pull request closed without merging**